snorkel/contrib/deps/test/labeling/model/test_dep_learner.py

```diff
@@ -0,0 +1,246 @@
+
+        # Choose random weights for the edges
+        te_min, te_max = min(theta_edge_range), max(theta_edge_range)
+        for (i, j) in self.E:
+            w_ij = (te_max - te_min) * random() + te_min
+            self.theta[(i, j)] = w_ij
+            self.theta[(j, i)] = w_ij
+
+    def _P(self, i, li, j, lj, y):
+        return np.exp(
+            self.theta[i][y] * indpm(li, y) + self.theta[(i, j)] * indpm(li, lj)
+        )
+
+    def P_conditional(self, i, li, j, lj, y):
+        """Compute the conditional probability
+            P_\theta(li | lj, y)
+            =
+            Z^{-1} exp(
+                theta_{i|y} \indpm{ \lambda_i = Y }
+                + \theta_{i,j} \indpm{ \lambda_i = \lambda_j }
+            )
+        In other words, compute the conditional probability that LF i outputs
+        li given that LF j output lj, and Y = y, parameterized by
+            - a class-conditional LF accuracy parameter \theta_{i|y}
+            - a symmetric LF correlation paramter \theta_{i,j}
+        """
+        Z = np.sum([self._P(i, _li, j, lj, y) for _li in range(self.k + 1)])
+        return self._P(i, li, j, lj, y) / Z
+
+    def _generate_label_matrix(self):
+        """Generate an [n,m] label matrix with entries in {0,...,k}"""
+        self.L = np.zeros((self.n, self.m))
+        self.Y = np.zeros(self.n, dtype=np.int64)
+        for i in range(self.n):
+            y = choice(self.k, p=self.p) + 1  # Note that y \in {1,...,k}
+            self.Y[i] = y
+            for j in range(self.m):
+                p_j = self.parent.get(j, 0)
+                prob_y = self.P_conditional(j, y, p_j, self.L[i, p_j], y)
+                prob_0 = self.P_conditional(j, 0, p_j, self.L[i, p_j], y)
+                p = np.ones(self.k + 1) * (1 - prob_y - prob_0) / (self.k - 1)
+                p[0] = prob_0
+                p[y] = prob_y
+                self.L[i, j] = choice(self.k + 1, p=p)
+
+    def _get_conditional_probs(self):
+        """Compute the true clique conditional probabilities P(\lC | Y) by
+        counting given L, Y; we'll use this as ground truth to compare to.
+
+        Note that this generates an attribute, self.c_probs, that has the same
+        definition as returned by `LabelModel.get_conditional_probs`.
+
+        TODO: Can compute these exactly if we want to implement that.
+        """
+        # TODO: Extend to higher-order cliques again
+        self.c_probs = np.zeros((self.m * (self.k + 1), self.k))
+        for y in range(1, self.k + 1):
+            Ly = self.L[self.Y == y]
+            for ly in range(self.k + 1):
+                self.c_probs[ly :: (self.k + 1), y - 1] = (
+                    np.where(Ly == ly, 1, 0).sum(axis=0) / Ly.shape[0]
+                )
+
+
+class DependencyLearnerTest(unittest.TestCase):
+    def test_cardinality(self):
+        dep_learner = DependencyLearner()
+        L = np.array([[-1, 0, 2], [1, 2, 3], [-1, -1, 0]])
```

Comment:
Nit: could use even more minimal test here for clarity; e.g., a 2x2 matrix that has -1, 0, 1, 2 or something.